clippy_lints/src/redundant_local.rs

```diff
@@ -0,0 +1,136 @@
+use clippy_utils::{diagnostics::span_lint_and_help, source::snippet_opt};
+use rustc_ast::{
+    ast::{BindingAnnotation, Block, Expr, ExprKind, LocalKind, Mutability, Pat, PatKind, Path, StmtKind},
+    visit::FnKind,
+    Local,
+};
+use rustc_data_structures::fx::FxHashMap;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{symbol::Ident, Span};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for redundant redefinitions of local bindings.
+    ///
+    /// ### Why is this bad?
+    /// Redundant redefinitions of local bindings do nothing and are likely to be unintended.
```

Comment:
I think this could also mention that it may allocate a new local on the stack (*even in release mode*, at least for `Vec<T>`. [example](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=ddcfaadc19ad7ce8497bcec03ec8d205))